br/pkg/stream/stream_metas.go

```diff
@@ -408,3 +321,947 @@ func ReplaceMetadata(meta *backuppb.Metadata, filegroups []*backuppb.DataFileGro
+				return
+			}
+
+			// Note: maybe make this a static method or just a normal function...
+			meta, err := (*MetadataHelper).ParseToMetadataHard(nil, data)
+			if err != nil {
+				emitErr(cannotBeRetryByRerunBase(
+					errors.Annotatef(err, "failed to parse meta %s", path)))
+				return
+			}
+
+			me := new(pb.MetaEdit)
+			me.Path = path
+			for _, ds := range meta.FileGroups {
+				if ds.MaxTs < from {
+					me.DeletePhysicalFiles = append(me.DeletePhysicalFiles, ds.Path)
+				}
+			}
+
+			err = m.applyMetaEditTo(ctx, me, meta)
+			if err != nil {
+				updateResult(func(r *MigratedTo) {
+					r.Warnings = append(r.Warnings, errors.Annotatef(err, "during handling %s", me.Path))
+					r.NewBase.EditMeta = append(r.NewBase.EditMeta, me)
+				})
+			}
+			m.cleanUpFor(ctx, me, out)
+			m.Hooks.DeletedAFileForTruncating(len(me.DeletePhysicalFiles))
+		})
+	}
+	wg.Wait()
+}
+
+type hookedStorage struct {
+	storage.ExternalStorage
+	metaSet *StreamMetadataSet
+}
+
+func (h hookedStorage) WriteFile(ctx context.Context, name string, data []byte) error {
+	if h.metaSet.BeforeDoWriteBack != nil {
+		meta, err := h.metaSet.Helper.ParseToMetadataHard(data)
+		if err != nil {
+			// Note: will this be too strict? But for now it seems this check won't fail.
+			// We can remove this in the future if needed.
+			return errors.Annotatef(err, "Writing non-meta during write back (to = %s)", name)
+		}
+		if h.metaSet.BeforeDoWriteBack(name, meta) {
+			log.Info("Skipped writeback meta by the hook.", zap.String("meta", name))
+			return nil
+		}
+	}
+
+	return h.ExternalStorage.WriteFile(ctx, name, data)
+}
+
+func (ms *StreamMetadataSet) hook(s storage.ExternalStorage) hookedStorage {
+	hooked := hookedStorage{
+		ExternalStorage: s,
+		metaSet:         ms,
+	}
+	if ms.DryRun {
+		hooked.ExternalStorage = storage.Batch(hooked.ExternalStorage)
+	}
+	return hooked
+}
+
+func physicalFileCanBeDeleted(fs *pb.DeleteSpansOfFile) bool {
+	sort.Slice(fs.Spans, func(i, j int) bool {
+		return fs.Spans[i].Offset < fs.Spans[j].Offset
+	})
+	lastOffset := uint64(0)
+	for _, s := range fs.Spans {
+		if s.Offset != lastOffset {
+			return false
+		}
+		lastOffset += s.Length
+	}
+	return lastOffset == fs.WholeFileLength
+}
+
+// mergeMetaEdits merges two meta edits.
+//
+// If the spans in the `DeleteLogicalFiles` consist a physical file,
+// they will be transformed to `DeletePhysicalFiles`.
+func mergeMetaEdits(s1, s2 []*pb.MetaEdit) []*pb.MetaEdit {
+	edits := map[string]*pb.MetaEdit{}
+	for _, edit := range s1 {
+		edits[edit.GetPath()] = &pb.MetaEdit{
+			Path:                edit.Path,
+			DeletePhysicalFiles: edit.DeletePhysicalFiles[:len(edit.DeletePhysicalFiles):len(edit.DeletePhysicalFiles)],
+			DeleteLogicalFiles:  edit.DeleteLogicalFiles[:len(edit.DeleteLogicalFiles):len(edit.DeleteLogicalFiles)],
+		}
+	}
+	for _, edit := range s2 {
+		target, ok := edits[edit.GetPath()]
+		if !ok {
+			edits[edit.GetPath()] = edit
```

Comment:
Should we also create a new `&pb.MetaEdit` the same as `edit` in `s1`?

Reply:
I guess no, we are copying the first one because we may modify the content of `Delete[Physical|Logical]Files` then. In this case we won't edit them.